src/Logger-PREVAC_TM1x/main.py

```diff
@@ -0,0 +1,370 @@
+
+        self.material_density = float(parameter["Material density in g/cm^3"])
+        self.impedance_ratio = float(parameter["Impedance ratio"])
+        self.tooling_factor = float(parameter["Tooling factor"])
+
+    def initialize(self) -> None:
+        """Get initial frequency."""
+        self.reset_thickness()
+        if self.initial_frequency == 0:
+            msg = "PREVAC TM1x: Initial frequency is 0. Check the connection to crystal head."
+            raise Exception(msg)
+
+    def call(self) -> (float, float):
+        """Return the current frequency."""
+        frequency = self.get_frequency()
+
+        thickness = self.calculate_thickness(
+            frequency,
+            self.initial_frequency,
+            self.material_density,
+            self.impedance_ratio,
+        )
+        thickness *= self.tooling_factor
+
+        rate = self.calculate_rate(thickness)
+        return [frequency, thickness, rate]
+
+    """ Device Functions """
+
+    def get_frequency(self) -> float:
+        """Request current frequency in Hz.
+
+        response[:4]    frequency in Hz with 0.01 Hz resolution
+        response[4]     1 microbalance correction connected, 2 not
+        response[5]     measurements are numbered from 0 to 255
+        response[6:7]   duration of the measurement in ms
+        response[7]     13 for TM13, 14 for TM14
+        """
+        command = 0x53
+        data = [self.sample_rate, 0, 0, 0]
+
+        self.send_dataframe(command, data)
+        response = self.get_dataframe()
+
+        frequency_bytes = response[:4].encode("latin-1")
+        frequency = int.from_bytes(frequency_bytes, "big")
+
+        return frequency / 100
+
+    @staticmethod
+    def calculate_thickness(
+        frequency: float,
+        initial_frequency: float,
+        density_material: float,
+        impedance_ratio: float,
+    ) -> float:
+        """Calculate the thickness from the frequency using Sauerbreys equation."""
+        freq_constant = 1.668e5  # Hz * cm
+        density_quartz = 2.648  # g/cm^3
+
+        normalization = (
+            density_quartz * freq_constant / (np.pi * density_material * impedance_ratio * initial_frequency)
+        )
+        ratio = np.arctan(impedance_ratio * np.tan(np.pi * (frequency - initial_frequency) / frequency))
+
+        # TODO: Check unit
+        # Convert from cm to nm
+        return normalization * ratio * 1e7 * -1
+
+    def reset_thickness(self) -> None:
+        """Reset thickness by updating initial frequency."""
+        self.initial_frequency = self.get_frequency()
+
+    def calculate_rate(self, thickness: float) -> float:
+        """Calculate the rate of the thickness change in A/s."""
+        time_stamp = time.time()
+        if self.previous_thickness == 0 or self.previous_time == 0:
+            rate = 0
+        else:
+            rate = (thickness - self.previous_thickness) / (time_stamp - self.previous_time)
```

Comment:
This could throw a division by zero error, if calculate rate is called more frequent than the system's time quantization.
In that case one could return the rate that was previously calculated (if available).

Same fallback could be applied if the numerator is zero and the function was called within a single sampling interval (if this is known).